File: kits/java/simple/lux/Agent.java
package lux;

import java.util.Scanner;

public class Agent {
    private final Scanner scanner;

    public GameState gameState = new GameState();

    /**
     * Constructor for a new agent User should edit this according to their `Design`
     */
    public Agent() {
      scanner = new Scanner(System.in);
    }

    /**
     * Initialize Agent for the `Match` User should edit this according to their
     * `Design`
     */
    public void initialize() {
      // get agent ID
      gameState.id = Integer.parseInt(scanner.nextLine());
      String mapInfo = scanner.nextLine();
      String[] mapInfoSplit = mapInfo.split(" ");
      int mapWidth = Integer.parseInt(mapInfoSplit[0]);
      int mapHeight = Integer.parseInt(mapInfoSplit[1]);
      gameState.map = new GameMap(mapWidth, mapHeight);
    }

    /**
     * Updates agent's own known state of `Match` User should edit this according to
     * their `Design`.
     */
    public void update() {
      // wait for the engine to send any updates
      gameState.map = new GameMap(gameState.map.width, gameState.map.height);
      gameState.turn += 1;
      gameState.players[0].cities.clear();
      gameState.players[0].units.clear();
      gameState.players[1].cities.clear();
      gameState.players[1].units.clear();
      while (true) {
        final String updateInfo = scanner.nextLine();
        if (updateInfo.equals(IOConstants.DONE.str)) {
          break;
        }
        String[] updates = updateInfo.split(" ");
        String inputIdentifier = updates[0];
        if (inputIdentifier.equals(IOConstants.RESEARCH_POINTS.str)) {
          int team = Integer.parseInt(updates[1]);
          this.gameState.players[team].researchPoints = Integer.parseInt(updates[2]);
        } else if (inputIdentifier.equals(IOConstants.RESOURCES.str)) {
          String type = updates[1];
          int x = Integer.parseInt(updates[2]);
          int y = Integer.parseInt(updates[3]);
          int amt = (int)(Double.parseDouble(updates[4]));
          this.gameState.map._setResource(type, x, y, amt);
        } else if (inputIdentifier.equals(IOConstants.UNITS.str)) {
          int i = 1;
          int unittype = Integer.parseInt(updates[i++]);
          int team = Integer.parseInt(updates[i++]);
          String unitid = updates[i++];
          int x = Integer.parseInt(updates[i++]);
          int y = Integer.parseInt(updates[i++]);
          double cooldown = Double.parseDouble(updates[i++]);
          int wood = Integer.parseInt(updates[i++]);
          int coal = Integer.parseInt(updates[i++]);
          int uranium = Integer.parseInt(updates[i++]);
          Unit unit = new Unit(team, unittype, unitid, x, y, cooldown, wood, coal, uranium);
          this.gameState.players[team].units.add(unit);
        } else if (inputIdentifier.equals(IOConstants.CITY.str)) {
          int i = 1;
          int team = Integer.parseInt(updates[i++]);
          String cityid = updates[i++];
          double fuel = Double.parseDouble(updates[i++]);
          double lightUpkeep = Double.parseDouble(updates[i++]);
          this.gameState.players[team].cities.put(cityid, new City(team, cityid, fuel, lightUpkeep));
        } else if (inputIdentifier.equals(IOConstants.CITY_TILES.str)) {
          int i = 1;
          int team = Integer.parseInt(updates[i++]);
          String cityid = updates[i++];
          int x = Integer.parseInt(updates[i++]);
          int y = Integer.parseInt(updates[i++]);
          double cooldown = Double.parseDouble(updates[i++]);
          City city = this.gameState.players[team].cities.get(cityid);
          CityTile citytile = city._add_city_tile(x, y, cooldown);
          this.gameState.map.getCell(x, y).citytile = citytile;
          this.gameState.players[team].cityTileCount += 1;
      } else if (inputIdentifier.equals(IOConstants.ROADS.str)) {
          int i = 1;
          int x = Integer.parseInt(updates[i++]);
          int y = Integer.parseInt(updates[i++]);
          double road = Double.parseDouble(updates[i++]);
          Cell cell = this.gameState.map.getCell(x, y);
          cell.road = road;
        }

      }
    }
    /**
     * End a turn
     */
    public void endTurn() {
        System.out.println("D_FINISH");
        System.out.flush();
    }
}

File: kits/ts/simple/lux/Agent.ts
import readline from "readline";
import {Player} from "./Player";
import {GameMap} from "./GameMap";
import {City} from "./City";
import {Unit} from "./Unit";
import {Parser} from "./Parser";
import {Parsed} from "./Parsed";
import {INPUT_CONSTANTS} from './io';

// Create parser and use ',' as the delimiter between commands being sent by the `Match` and `MatchEngine`
const parse = new Parser(' ');

export interface GameState {
  id: number;
  map: GameMap;
  players: Array<Player>;
  turn: number;
}

/**
 * Agent for sequential `Designs`
 */
export class Agent {
  public getLine: () => Promise<Parsed>;
  public gameState: GameState;
  public _setup(): void {

    // Prepare to read input
    const rl = readline.createInterface({
      input: process.stdin,
      output: null,
    });

    const buffer = [];
    let currentResolve: () => void;
    let currentPromise;
    const makePromise = function () {
      return new Promise<void>((resolve) => {
        currentResolve = resolve;
      });
    };
    // on each line, push line to buffer
    rl.on('line', (line) => {
      buffer.push(line);
      currentResolve();
      currentPromise = makePromise();
    });
    // The current promise for retrieving the next line
    currentPromise = makePromise()


    // with await, we pause process until there is input
    this.getLine = async () => {
      return new Promise(async (resolve) => {
        while (buffer.length === 0) {
          // pause while buffer is empty, continue if new line read
          await currentPromise;
        }
        // once buffer is not empty, resolve the most recent line in stdin, and remove it
        resolve((parse as any)(buffer.shift()));
      });
    };
  }

  /**
   * Constructor for a new agent
   * User should edit this according to the `Design` this agent will compete under
   */
  public constructor() {
    this._setup(); // DO NOT REMOVE
  }

  /**
   * Initialize Agent for the `Match`
   * User should edit this according to their `Design`
   */
  public async initialize() {

    // get agent ID
    const id = (await this.getLine()).nextInt();
    // get some other necessary initial input
    let mapInfo = (await this.getLine());
    let width = mapInfo.nextInt();
    let height = mapInfo.nextInt();
    const map = new GameMap(width, height);
    const players = [new Player(0), new Player(1)];

    this.gameState = {
      id,
      map,
      players,
      turn: -1,
    };
  }
  /**
   * Updates agent's own known state of `Match`
   * User should edit this according to their `Design`.
   */
  public async update() {
    this.gameState.turn++;
    // wait for the engine to send any updates
    await this.retrieveUpdates();
  }

  public resetPlayerStates() {
    const players = this.gameState.players;
    players[0].units = [];
    players[0].cities = new Map();
    players[0].cityTileCount = 0;
    players[1].units = [];
    players[1].cities = new Map();
    players[1].cityTileCount = 0;
  }
  public async retrieveUpdates() {
    this.resetPlayerStates();
    // TODO: this can be optimized. we only reset because some resources get removed
    this.gameState.map = new GameMap(this.gameState.map.width, this.gameState.map.height);
    while (true) {
      let update = (await this.getLine());
      if (update.str === INPUT_CONSTANTS.DONE) {
        break;
      }
      const inputIdentifier = update.nextStr();
      switch (inputIdentifier) {
        case INPUT_CONSTANTS.RESEARCH_POINTS: {
          const team = update.nextInt();
          this.gameState.players[team].researchPoints = update.nextInt();
          break;
        }
        case INPUT_CONSTANTS.RESOURCES: {
          const type = update.nextStr();
          const x = update.nextInt();
          const y = update.nextInt();
          const amt = update.nextInt();
          this.gameState.map._setResource(type, x, y, amt);
          break;
        }
        case INPUT_CONSTANTS.UNITS: {
          const unittype = update.nextInt();
          const team = update.nextInt();
          const unitid = update.nextStr();
          const x = update.nextInt();
          const y = update.nextInt();
          const cooldown = update.nextFloat();
          const wood = update.nextInt();
          const coal = update.nextInt();
          const uranium = update.nextInt();
          this.gameState.players[team].units.push(new Unit(team, unittype, unitid, x, y, cooldown, wood, coal, uranium));
          break;
        }
        case INPUT_CONSTANTS.CITY: {
          const team = update.nextInt();
          const cityid = update.nextStr();
          const fuel = update.nextFloat();
          const lightUpkeep = update.nextFloat();
          this.gameState.players[team].cities.set(cityid, new City(team, cityid, fuel, lightUpkeep));
          break;
        }
        case INPUT_CONSTANTS.CITY_TILES: {
          const team = update.nextInt();
          const cityid = update.nextStr();
          const x = update.nextInt();
          const y = update.nextInt();
          const cooldown = update.nextFloat();
          const city = this.gameState.players[team].cities.get(cityid);
          const citytile = city.addCityTile(x, y, cooldown);
          this.gameState.map.getCell(x, y).citytile = citytile;
          this.gameState.players[team].cityTileCount += 1;
          break;
        }
        case INPUT_CONSTANTS.ROADS: {
          const x = update.nextInt();
          const y = update.nextInt();
          const road = update.nextFloat();
          this.gameState.map.getCell(x, y).road = road;
          break;
        }
      }
    }
  }

  /**
   * End a turn
   */
  public endTurn(): void {
    console.log('D_FINISH');
  }

  public async run(loop: (gameState: GameState) => Array<string>): Promise<void> {
    await this.initialize();
    while (true) {
      await this.update();
      try {
        const actions = loop(this.gameState);
        console.log(actions.join(","));
      } catch (err) {
        console.log(err);
      }
      this.endTurn();
    }
  }
}

export const annotate = {
  circle: (x: number, y: number): string => {
    return `dc ${x} ${y}`
  },
  x: (x: number, y: number): string => {
    return `dx ${x} ${y}`
  },
  line: (x1: number, y1: number, x2: number, y2: number): string => {
    return `dl ${x1} ${y1} ${x2} ${y2}`
  },
  text: (x1: number, y1: number, message: string, fontsize: number = 16) => {
    return `dt ${x1} ${y1} '${message}' ${fontsize}`
  },
  sidetext: (message: string) => {
    return `dst '${message}'`
  }
}
